src/mem/ruby/protocol/GPU_VIPER-TCC.sm

```diff
@@ -707,6 +739,10 @@ machine(MachineType:TCC, "TCC Cache")
     }
   }
 
+  action(dnant_decrementNumAtomicsNoTrigger, "dnant", desc="inc num atomics") {
```

Comment:
What is the "no trigger" part about?

Reply:
The trigger queue appears to be in place to account for race conditions where an atomic request that gets performed in the directory is sent to the directory while another atomic request that gets sent to the directory is being processed by the directory.

In cases where the atomic is performed in the directory I don't believe this race condition exists since the directory is only reading data; the behavior would be similar to reads that go to the directory, which don't have a trigger queue. Because of this, and because when the trigger queue is emptied the atomic is transitioned to invalid, I've circumvented the trigger queue by adding a way to directly decrement the atomic count without sending a trigger. This is only used in a WB L2 with atomic requests that are performed locally on the return path.

Reply:
I don't think we should circumvent the trigger queue.  Doing this in the past is how we inadvertently introduced races into this protocol (e.g., because of replays when the queue was full).

Reply:
The trigger queue in the previous case was used in a transition that isn't guaranteed to change states, so it used the trigger queue to change the state iif the atomic count was zero.  I think this case is different because we will always transition from A>M on data and then I believe further atomics will stay M>M?  Will let Dan confirm though.

Reply:
@DanKouch looks like this name still needs to be updated?

Reply:
@mattsinc Yep, just rebased for now. I have the changes ready but I'm running into some trouble with non-returning atomics. I'll push as soon as I get that sorted.